---
title: "1.Clustering, Peak calling, re-processing B-cell dataset, DARs and DEG analysis"
author: 
- name: "Maria Lucia Romero Rivero"
  email: MLROMEROR@clinic.cat
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
    fig_captions: yes

---

# Introduction

Hi to everyone!
In this notebook we are going to perform the clustering to localize the  B-cell cluster base on GEx. Once we detect the B-cell specific clusters we are going to subset these cells and re process them, to do so we will need to call the peaks of these new data set, since the peaks may not be the same than we got using the whole dataset. After that we will integrate both techniques by using WNN. Then we will perform both DARs and DEG analysis and we will save them for filtering the Peak-gene links in the last notebook. 


<details>
  <summary>**Load packages**</summary>
  

```{r eval=FALSE}
BiocManager::install("Repitools")

```

```{r setup, message=FALSE, warning=FALSE}
library(Signac)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(tidyverse)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(plyr)
library(reshape2)
library(data.table)
library(GenomicRanges)
library(harmony)
library(hdf5r)
library(stringr)
library(RColorBrewer)
library(magick)
library(knitr) 
library(biovizBase)
library(patchwork)
library(Repitools)
library(kableExtra)
set.seed(123)
```

</details>

# Loading data
 
## Directory path 

```{r}
# Paths
path_to_data <- here::here("results/R_objects/")
path_to_save <- here::here("results/R_objects/")
path_to_save_tables <- here::here("results/tables/")

```


```{r message=FALSE, warning=FALSE}
tonsil_normalized <- readRDS(paste0(path_to_data,"2.tonsil_merged_harmony.rds"))
```



# Clustering

**FindClusters**
Identify clusters of cells by a shared nearest neighbor (SNN) modularity optimization 
based clustering algorithm. First calculate k-nearest neighbors and construct 
the SNN graph. Then optimize the modularity function to determine clusters.

## FindClusters

```{r}
#find cluster algorithm 3 = SLM algorithm

tonsil_normalized <- FindClusters(tonsil_normalized, resolution = 0.02,algorithm = 3, graph.name = "wsnn",verbose = FALSE)
print(colnames(tonsil_normalized@meta.data))
```

## Joint UMAP

```{r fig.wide=TRUE}
 DimPlot(
    tonsil_normalized,
    group.by =  "wsnn_res.0.02",
    reduction = "wnn.umap",
    pt.size = 0.1, label = T
  )
```

## RNA-seq

```{r fig.wide=TRUE}
 DimPlot(
    tonsil_normalized,
    group.by =  "wsnn_res.0.02",
    reduction = "umap.rna",
    pt.size = 0.1, label = T
  )
```


## ATAC-seq

```{r fig.wide=TRUE}
 DimPlot(
    tonsil_normalized,
    group.by =  "wsnn_res.0.02",
    reduction = "umap.atac",
    pt.size = 0.1, label = T
  )
```

## Dotplot of GEX markers 

In this case is important to set the RNA assay as a default assay otherwise you won't be able to get dene expression plot.

```{r}
DefaultAssay(tonsil_normalized)<-"RNA"
```


```{r}
m<-c("SUGCT","BANK1", "CXCR4", "MS4A1","RALGPS2","CD83","AICDA","BCL2A1","MEF2B", "BCL6","XBP1","IRF4","PRDM1","SLAMF7", "MZB1", "FKBP11","NKG7", "CD3D", "IL7R","CD8A","CD2","LEF1","LYZ", "S100A8")

```
```{r}
dot<-DotPlot(tonsil_normalized, features = m,cols = c("lightgrey","blue"), cluster.idents = T) + theme(axis.text.x = element_text( size = 10, vjust = 0.8, hjust = 0.8)) +ggtitle("res 0.02 markers of each cluster")
dot+
  coord_flip() +
  theme(axis.title = element_blank(), axis.text.y = element_text(size = 7))
```

## Rename clusters

```{r}
new.cluster.ids <- c("NBC_MBC","GCBC", "T-cells", "PC", "Myeloid")
names(new.cluster.ids) <- levels(tonsil_normalized)
tonsil_normalized <- RenameIdents(tonsil_normalized, new.cluster.ids) 
DimPlot(tonsil_normalized, reduction = "wnn.umap", label = TRUE, pt.size = 0.5)+NoLegend()
```
```{r}
tonsil_normalized$seurat_clusters_name<-Idents(tonsil_normalized)

```

# Subset B-cell clusters

Once we localize the main B-cell types. We will subset the specific B-cell cluster to work with only these data. 

```{r}

tonsil_normalized$is_bcell <- 
  tonsil_normalized$seurat_clusters_name == "GCBC" |
  tonsil_normalized$seurat_clusters_name == "NBC_MBC" |
  tonsil_normalized$seurat_clusters_name == "PC" 

tonsil_bcell <- subset(tonsil_normalized, subset = is_bcell == TRUE)
```


Now, we check if we are subseting the interesting clusters. 

```{r}
DimPlot(tonsil_bcell, reduction = "wnn.umap", label = TRUE, pt.size = 0.5)+NoLegend()
```


Once we subset the target clusters we have to normalize again the data and performe the Peak Calling process 

# Peak calling 

After subset the data we want to analyse, we have to perform the peak calling process again to re organize the peaks to the new set of peaks 

```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"

```

*Peak Calling using MACS2*

First of all we have to download in the terminar the [MACS2](https://github.com/macs3-project/MACS) package (by conda, pypi etc) you have to take into account that MACS v3.x.x requires Python3. They have tested MACS in Python3.6, 3.7 and 3.8. [Instalation](https://github.com/macs3-project/MACS/blob/master/docs/INSTALL.md)

CalPeaks using MACS. Fragment files linked to the specified assay will be used to call peaks. If multiple fragment files are present, all will be used in a single MACS invocation. Returns the .narrowPeak MACS output as a GRanges object.

It can takes arounf 12-15 min 

*S3 method for class 'Seurat'*

* `CallPeaks`(

  `seurat_object`,
  
  `group.by` = Grouping variable to use. If set, peaks will be called independently on each group of cells and then combined. Note that to call peaks using subsets of cells we first split the fragment file/s used, so using a grouping variable will require extra time to split the files and perform multiple MACS peak calls, and will store additional files on-disk that may be large. Note that we store split fragment files in the temp directory (tempdir) by default, and if the program is interrupted before completing these temporary files will not be removed. If NULL, peaks are called using all cells together (pseudobulk).,

  `macs2.path` = Path to MACS program. If NULL, try to find MACS automatically.
  )


```{r message=FALSE, warning=FALSE}
peaks <- CallPeaks(tonsil_bcell, macs2.path = "/Users/mlromeror/opt/anaconda3/envs/myenvpython2/bin/macs2",group.by="wsnn_res.0.02")
saveRDS(peaks, "results/tables/2.MACS_annotation_level_1.rds")
```

* `annoGR2DF`: Convert an annotated GRanges to a data.frame.

```{r}

df <- annoGR2DF(peaks)
write.csv(df, "results/tables/2.df_MACS_annotation_level_1.csv")
```

If we visualize the data frame of the peak calling, we can observe that it has a GRanger format.
* `chr`: reference chromosome
* `start`: start coordinate of the peak
* `end`: end coordinate of the peak
* `width`: end-start. number of bp in each peak
* `peak_called_in`: in which cluster was found 
 
For instance peak 5 was found in cluster 0, 1 and 3


```{r}
head(df)
```

Remove peaks on nonstandard chromosomes and in genomic blacklist regions

```{r}
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)
saveRDS(peaks, "results/tables/2.peaks_annotation_level_1_subset.rds")
```

## Quantify counts in each peak:

Construct a feature x cell matrix from a genomic fragments file. It can take 16 min

```{r}
macs2_counts <- FeatureMatrix(
  fragments = Fragments(tonsil_bcell),
  features = peaks,
  cells = colnames(tonsil_bcell)
)
```


**create a new assay using the MACS2 peak set and add it to the Seurat object**


```{r message=FALSE, warning=FALSE}
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
ensembldb::seqlevelsStyle(annotation) <- "UCSC"

```

Save fragment path in the `fragment` object 

```{r}
#fragment directory path 
fragments<-Fragments(tonsil_bcell[["ATAC"]])

#Crate a new Chromatin assay with the new peek calling 
tonsil_bcell[["ATAC"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragments,
  annotation = annotation
)
```


# Normalization and lineal and non-lineal dimentionallity reduction (LSI and UMAP)

## scATAC

RunSVD LSI
```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"
tonsil_bcell <- RunTFIDF(tonsil_bcell)
tonsil_bcell <- FindTopFeatures(tonsil_bcell, min.cutoff = "q0")
tonsil_bcell <- RunSVD(tonsil_bcell)
```


### Plot the Depth correlation plot

For scRNA-seq data we don't typically observe such a strong relationship between the first PC and sequencing depth, and so usually retain the first PC in downstream analyses.

```{r fig.wide=TRUE}

DepthCor(tonsil_bcell)

```

Here we see there is a very strong correlation between the first LSI component and the total number of counts for the cell, so we will perform downstream steps without this component R~-1


### UMAP representation

```{r fig.wide=TRUE}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "lsi",
  reduction.name = "umap.atac",
  reduction.key = "atacUMAP_"
)
atac.umap<-DimPlot(
  tonsil_bcell,
  reduction = "umap.atac",
  group.by = "library_name",
  pt.size = 0.1
) + ggtitle('scATAC UMAP') + NoLegend()
```
   
   
```{r}
atac.umap 
```


## scRNA-seq

Our aim is to detect and exclude empty droplets or death cells (lysed cells). Lysed cells have 3 hallmarks:
- (1) low library size (total UMI), 
- (2) low library complexity (number of detected genes) a
- (3) high fraction of mitochondrial expression (cytosolic mRNA leaks out of the cell). 

<details>
  <summary>**scRNA-seq**</summary>


### Normalization and linear dimensional reduction- 

### NormalizeData (Log Normalization)

```{r message=FALSE, warning=FALSE}
DefaultAssay(tonsil_bcell) <- "RNA"
tonsil_bcell <- NormalizeData(
  tonsil_bcell,
  normalization.method = "LogNormalize",
  scale.factor = 1e4
)

tonsil_bcell <- tonsil_bcell %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>% 
  RunPCA() 
```



### UMAP representation

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 1:30,
  reduction = "pca",
  reduction.name = "umap.rna",
  reduction.key = "rnaUMAP_"
)
rna.umap<-DimPlot(
  tonsil_bcell,
  reduction = "umap.rna",
  group.by = "library_name",
  pt.size = 0.1) + NoLegend() + ggtitle('scRNA UMAP')

rna.umap
```

 

```{r}
atac.umap + rna.umap
```

# Harmony Integration (correct data by bath effect)

Pass the Seurat object to the RunHarmony function and specify which variable to integrate 
out. A Seurat object is returned with corrected Harmony coordinates.

## scATAC

```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"
tonsil_bcell <- RunHarmony(
  object = tonsil_bcell,
  reduction = "lsi",
  dims = 2:40,
  group.by.vars = "library_name",
  assay.use = "ATAC",
  project.dim = FALSE,
  reduction.save = "harmony_atac"
)
```


### UMAP representation

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "harmony_atac",
  reduction.name = "umap.atac",
  reduction.key = "atacUMAP_"
)

Harm_peak<-DimPlot(
  tonsil_bcell,
  reduction = "umap.atac",
  group.by = "library_name",
  pt.size = 0.1
) + NoLegend() + ggtitle('ATAC Harmony')
```

## scRNA

<details>
  <summary>**harmony in RNA-seq**</summary>
  

```{r message=FALSE, warning=FALSE, include=FALSE}
DefaultAssay(tonsil_bcell) <- "RNA"
tonsil_bcell <- RunHarmony(
  object = tonsil_bcell,
  reduction = "pca",
  dims = 1:30,
  group.by.vars = "library_name",
  assay.use = "RNA",
  project.dim = FALSE,
  reduction.save = "harmony_rna"
)
```

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "harmony_rna",
  reduction.name = "umap.rna",
  reduction.key = "rnaUMAP_"
)
Harm_rna<-DimPlot(
  tonsil_bcell,
  reduction = "umap.rna",
  group.by = "library_name",
  pt.size = 0.1
) + NoLegend() + ggtitle('RNA Harmony')
```

 

## scATAC and RNAseq Harmony

```{r}
Harm_peak+Harm_rna + plot_annotation(title = 'Harmony ATAC and RNA UMAP visualization')
```

# Joint UMAP visualization

**FindModalNeighbors**

This function will construct a weighted nearest neighbor (WNN) graph. For each cell, we identify the nearest neighbors based on a weighted combination of two modalities. Takes as input two dimensional reductions, one computed for each modality. Other parameters are listed for debugging, but can be left as default values.

* *object*: A Seurat object
* *reduction.list*: A list of two dimensional reductions, one for each of the modalities to be integrated
* *dims.list*: A list containing the dimensions for each reduction to use
* *modality.weight.name*: Variable name to store modality weight in object meta data
* *verbose*: Print progress bars and output

```{r}
# build a joint neighbor graph using both assays
tonsil_bcell <- FindMultiModalNeighbors(
  object = tonsil_bcell,
  reduction.list = list("harmony_atac", "harmony_rna"),
  dims.list = list(2:40, 1:30), 
  modality.weight.name = "Joint_wnn_umap"
  )

```


Then we are going to build a joint UMAP visualization base on the 
 
```{r}
tonsil_bcell <- RunUMAP(
  object = tonsil_bcell,
  nn.name = "weighted.nn",
  reduction.name = "wnn.umap",
  reduction.key = "wnnUMAP_")

joint.umap<- DimPlot(tonsil_bcell, label = FALSE, group.by = "library_name", pt.size = 0.1,  reduction = "wnn.umap") + plot_annotation(title = 'Joint UMAP')+ ggtitle('Joint UMAP by library name') + NoLegend()

```

```{r}
joint.umap 
```


# Clustering

## FindClusters

```{r}
#find cluster algorithm 3 = SLM algorithm

tonsil_bcell <- FindClusters(tonsil_bcell, resolution = c(0.02,0.05,0.06),algorithm = 3, graph.name = "wsnn",verbose = FALSE)
print(colnames(tonsil_bcell@meta.data))
```
```{r fig.wide=TRUE}
vars <- str_subset(colnames(tonsil_bcell@meta.data), "^wsnn_res")
clusters_gg <- purrr::map(vars, function(x) {
  p <- DimPlot(
    tonsil_bcell,
    group.by = x,
    reduction = "wnn.umap",
    pt.size = 0.1, label = T
  )
  p 
})
clusters_gg
```



## Joint UMAP

```{r fig.wide=TRUE}
 DimPlot(
    tonsil_bcell,
    group.by =  "wsnn_res.0.06",
    reduction = "wnn.umap",
    pt.size = 0.1, label = T
  )
```



## Dotplot of GEX markers 

```{r}
DefaultAssay(tonsil_bcell)<-"RNA"
m<-c("FCER2","IGHD","IGHM", "SUGCT","BANK1", "CXCR4", "MS4A1","RALGPS2","CD83","AICDA","BCL2A1","MEF2B","RGS13","SOX5", "BCL6","XBP1","IRF4","PRDM1","SLAMF7", "MZB1","CD3E","CD4","CD8A","CD3D")

```
```{r}
Idents(tonsil_bcell)<-"wsnn_res.0.06"
dot<-DotPlot(tonsil_bcell, features = m,cols = c("lightgrey","blue"), cluster.idents = T) + theme(axis.text.x = element_text( size = 10, vjust = 0.8, hjust = 0.8)) +ggtitle("res 0.02 markers of each cluster")
dot+
  coord_flip() +
  theme(axis.title = element_blank(), axis.text.y = element_text(size = 7))
```

In this dotplot we can easily notice that the cluster 5 has a high expresion of T-cell and B-cell  genes so we can classify it as cluster of doublets cells. 

I also like to see the expression of those markers in the UMAP to see if the expression is unique in this cluster or in others.

# Feature plot function

```{r}
markers_featureplot <- function(x){purrr::map(x, function(x) {
  
  p <- FeaturePlot(
    tonsil_bcell,
    features = x,
    reduction = "wnn.umap",
    pt.size = 0.1
  )
  p
})}
```

## T-cell expression

```{r}
Bib_Tcell_markers<-c("CD3E","CD4","CD8A","CD3D")
```

```{r}
markers_featureplot(Bib_Tcell_markers)
```

# Remover doublet cluster

Since we found a cluster with doublet cells, we will subset the specific B-cell cluster and remover the doublet cells to work with high quality cells. 

```{r}

tonsil_bcell$is_cluster5 <- 
  tonsil_bcell$wsnn_res.0.06 == "5" 

tonsil_bcell <- subset(tonsil_bcell, subset = is_cluster5 == FALSE)
```


Now, we check if we are subseting the interesting clusters. 

```{r}
DimPlot(tonsil_bcell, reduction = "wnn.umap", label = TRUE, pt.size = 0.5)+NoLegend()
```


Once we subset the target clusters we have to normalize again the data and performe the Peak Calling process 

# Peak calling 

After subset the data we want to analyse, we have to perform the peak calling process again to re organize the peaks to the new set of peaks 

```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"

```

**Peak Calling using MACS2**

**S3 method for class 'Seurat'**

`CallPeaks`(
  `seurat_object`,
  
  `group.by` = Grouping variable to use. If set, peaks will be called independently on each group of cells and then combined. Note that to call peaks using subsets of cells we first split the fragment file/s used, so using a grouping variable will require extra time to split the files and perform multiple MACS peak calls, and will store additional files on-disk that may be large. Note that we store split fragment files in the temp directory (tempdir) by default, and if the program is interrupted before completing these temporary files will not be removed. If NULL, peaks are called using all cells together (pseudobulk).,

  `macs2.path` = Path to MACS program. If NULL, try to find MACS automatically.
  )


```{r message=FALSE, warning=FALSE}
peaks <- CallPeaks(tonsil_bcell, macs2.path = "/Users/mlromeror/opt/anaconda3/envs/myenvpython2/bin/macs2",group.by="wsnn_res.0.02")
saveRDS(peaks, "results/tables/2.MACS_annotation_level_1.2.rds")
```

`annoGR2DF`: Convert an annotated GRanges to a data.frame.

```{r}

df <- annoGR2DF(peaks)
write.csv(df, "results/tables/2.df_MACS_annotation_level_1.2.csv")
```

If we visualize the data frame of the peak calling, we can observe that it has a GRanger format.
`chr`: reference chromosome
`start`: start coordinate of the peak
`end`: end coordinate of the peak
`width`: end-start. number of bp in each peak
`peak_called_in`: in which cluster was found 
 
For instance peak 5 was found in cluster 0, 1 and 2


```{r}
head(df)
```

Remove peaks on nonstandard chromosomes and in genomic blacklist regions

```{r}
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)
saveRDS(peaks, "results/tables/2.peaks_annotation_level_1_subset.rds")
```

## Quantify counts in each peak:

Construct a feature x cell matrix from a genomic fragments file. It can take 16 min

```{r}
macs2_counts <- FeatureMatrix(
  fragments = Fragments(tonsil_bcell),
  features = peaks,
  cells = colnames(tonsil_bcell)
)
```


**create a new assay using the MACS2 peak set and add it to the Seurat object**


```{r message=FALSE, warning=FALSE}
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
ensembldb::seqlevelsStyle(annotation) <- "UCSC"

```

Save fragment path in the `fragments` object 

```{r}
#fragment directory path 
fragments<-Fragments(tonsil_bcell[["ATAC"]])

#Crate a new Chromatin assay with the new peek calling 
tonsil_bcell[["ATAC"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragments,
  annotation = annotation
)
```


# Normalization and lineal and non-lineal dimentionallity reduction (LSI and UMAP)

## scATAC

RunSVD LSI
```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"
tonsil_bcell <- RunTFIDF(tonsil_bcell)
tonsil_bcell <- FindTopFeatures(tonsil_bcell, min.cutoff = "q0")
tonsil_bcell <- RunSVD(tonsil_bcell)
```


### Plot the Depth correlation plot

For scRNA-seq data we don't typically observe such a strong relationship between the first PC and sequencing depth, and so usually retain the first PC in downstream analyses.

```{r fig.wide=TRUE}

DepthCor(tonsil_bcell)

```

Here we see there is a very strong correlation between the first LSI component and the total number of counts for the cell, so we will perform downstream steps without this component R~-1


### UMAP representation

```{r fig.wide=TRUE, message=FALSE, warning=FALSE}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "lsi",
  reduction.name = "umap.atac",
  reduction.key = "atacUMAP_"
)
atac.umap<-DimPlot(
  tonsil_bcell,
  reduction = "umap.atac",
  group.by = "library_name",
  pt.size = 0.1
) + ggtitle('scATAC UMAP') + NoLegend()
```
   
   
```{r}
atac.umap 
```


## scRNA-seq

Our aim is to detect and exclude empty droplets or death cells (lysed cells). Lysed cells have 3 hallmarks:
- (1) low library size (total UMI), 
- (2) low library complexity (number of detected genes) a
- (3) high fraction of mitochondrial expression (cytosolic mRNA leaks out of the cell). 

<details>
  <summary>**scRNA-seq**</summary>


### Normalization and linear dimensional reduction- 

### NormalizeData (Log Normalization)

```{r message=FALSE, warning=FALSE}
DefaultAssay(tonsil_bcell) <- "RNA"
tonsil_bcell <- NormalizeData(
  tonsil_bcell,
  normalization.method = "LogNormalize",
  scale.factor = 1e4
)

tonsil_bcell <- tonsil_bcell %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>% 
  RunPCA() 
```



### UMAP representation

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 1:30,
  reduction = "pca",
  reduction.name = "umap.rna",
  reduction.key = "rnaUMAP_"
)
rna.umap<-DimPlot(
  tonsil_bcell,
  reduction = "umap.rna",
  group.by = "library_name",
  pt.size = 0.1) + NoLegend() + ggtitle('scRNA UMAP')

rna.umap
```

 

```{r}
atac.umap + rna.umap
```

# Harmony Integration (correct data by bath effect)

Pass the Seurat object to the RunHarmony function and specify which variable to integrate 
out. A Seurat object is returned with corrected Harmony coordinates.

## scATAC

```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"
tonsil_bcell <- RunHarmony(
  object = tonsil_bcell,
  reduction = "lsi",
  dims = 2:40,
  group.by.vars = "library_name",
  assay.use = "ATAC",
  project.dim = FALSE,
  reduction.save = "harmony_atac"
)
```


### UMAP representation

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "harmony_atac",
  reduction.name = "umap.atac",
  reduction.key = "atacUMAP_"
)

Harm_peak<-DimPlot(
  tonsil_bcell,
  reduction = "umap.atac",
  group.by = "library_name",
  pt.size = 0.1
) + NoLegend() + ggtitle('ATAC Harmony')
```

## scRNA

<details>
  <summary>**harmony in RNA-seq**</summary>
  

```{r message=FALSE, warning=FALSE, include=FALSE}
DefaultAssay(tonsil_bcell) <- "RNA"
tonsil_bcell <- RunHarmony(
  object = tonsil_bcell,
  reduction = "pca",
  dims = 1:30,
  group.by.vars = "library_name",
  assay.use = "RNA",
  project.dim = FALSE,
  reduction.save = "harmony_rna"
)
```

```{r}
tonsil_bcell <- RunUMAP(
  tonsil_bcell,
  dims = 2:40,
  reduction = "harmony_rna",
  reduction.name = "umap.rna",
  reduction.key = "rnaUMAP_"
)
Harm_rna<-DimPlot(
  tonsil_bcell,
  reduction = "umap.rna",
  group.by = "library_name",
  pt.size = 0.1
) + NoLegend() + ggtitle('RNA Harmony')
```

 

## scATAC and RNAseq Harmony

```{r}
Harm_peak+Harm_rna + plot_annotation(title = 'Harmony ATAC and RNA UMAP visualization')
```

# Joint UMAP visualization

**FindModalNeighbors**

* **object**: A Seurat object
* **reduction.list**: A list of two dimensional reductions, one for each of the modalities to be integrated
* **dims.list**: A list containing the dimensions for each reduction to use
* **modality.weight.name**: Variable name to store modality weight in object meta data
* **verbose**: Print progress bars and output

```{r}
# build a joint neighbor graph using both assays
tonsil_bcell <- FindMultiModalNeighbors(
  object = tonsil_bcell,
  reduction.list = list("harmony_atac", "harmony_rna"),
  dims.list = list(2:40, 1:30), 
  modality.weight.name = "Joint_wnn_umap"
  )

```


```{r}
tonsil_bcell <- RunUMAP(
  object = tonsil_bcell,
  nn.name = "weighted.nn",
  reduction.name = "wnn.umap",
  reduction.key = "wnnUMAP_")

joint.umap<- DimPlot(tonsil_bcell, label = FALSE, group.by = "library_name", pt.size = 0.1,  reduction = "wnn.umap") + plot_annotation(title = 'Joint UMAP')+ ggtitle('Joint UMAP by library name') + NoLegend()

```

```{r}
joint.umap 
```


# Clustering

## FindClusters

```{r}
#find cluster algorithm 3 = SLM algorithm

tonsil_bcell <- FindClusters(tonsil_bcell, resolution = c(0.02,0.05,0.075),algorithm = 3, graph.name = "wsnn",verbose = FALSE)
print(colnames(tonsil_bcell@meta.data))
```



```{r fig.wide=TRUE}
vars <- str_subset(colnames(tonsil_bcell@meta.data), "^wsnn_res")
clusters_gg <- purrr::map(vars, function(x) {
  p <- DimPlot(
    tonsil_bcell,
    group.by = x,
    reduction = "wnn.umap",
    pt.size = 0.1, label = T
  )
  p 
})
clusters_gg
```

# Dotplot of GEX markers 

```{r}
DefaultAssay(tonsil_bcell)<-"RNA"
m<-c("FCER2","IGHD","IGHM", "SUGCT","BANK1", "CXCR4", "MS4A1","RALGPS2","CD83","AICDA","BCL2A1","MEF2B","RGS13","SOX5", "BCL6","XBP1","IRF4","PRDM1","SLAMF7", "MZB1","CD3E","CD4","CD8A","CD3D")

```


```{r}
Idents(tonsil_bcell)<-"wsnn_res.0.075"
dot<-DotPlot(tonsil_bcell, features = m,cols = c("lightgrey","blue"), cluster.idents = T) + theme(axis.text.x = element_text( size = 10, vjust = 0.8, hjust = 0.8)) +ggtitle("res 0.02 markers of each cluster")
dot+
  coord_flip() +
  theme(axis.title = element_blank(), axis.text.y = element_text(size = 7))
```



# Rename clusters

```{r}
new.cluster.ids <- c("GCBC", "NBC","GCBC", "MBC","PC")
names(new.cluster.ids) <- levels(tonsil_bcell)
tonsil_bcell <- RenameIdents(tonsil_bcell, new.cluster.ids) 

#save annotation 
tonsil_bcell[["annotation_level_1"]] <- Idents(object = tonsil_bcell)


DimPlot(tonsil_bcell,label = TRUE, reduction = "wnn.umap", pt.size = 0.5,cols = c("#a6cee3", "#1f78b4","#b2df8a", "#fb9a99","#e31a1c")) + ggtitle("Joint UMAP")+NoLegend()

```
```{r}
tonsil_bcell$seurat_clusters_name<-Idents(tonsil_bcell)

```

# Differentially expressed genes (DEG)

```{r}
DefaultAssay(tonsil_bcell)<-"RNA"
tonsil_bcell_markers<-FindAllMarkers(object = tonsil_bcell, only.pos = TRUE)
write.csv(tonsil_bcell_markers,file=paste0(path_to_save_tables, "tonsil_bcell_deg.csv"))
```

### Get top 10 markers of each cluster

**To 10 markers in resolution 0.02**
 
```{r}
top10_tonsil_bcell_markers<-tonsil_bcell_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
```

```{r}
df_top10<-as.data.frame(top10_tonsil_bcell_markers)
kbl(df_top10,caption = "Table of the top 5 marker of each cluster") %>%
  kable_paper("striped", full_width = F)
```


# Differentially Accessible Region (DARs)

Here we are going to to perform a differentially accessible region analysis, so that means that we are going to select all those region that are differentially open in one cluster compared to the rest. So in this case instead of obtaining a table with a column with differentially expressed genes, this column will be the coordinate of the differentially accesible region.


```{r}
DefaultAssay(tonsil_bcell) <- "ATAC"

tonsil_bcell_dars<-FindAllMarkers(object = tonsil_bcell, only.pos = TRUE)
write.csv(tonsil_bcell_dars,file=paste0(path_to_save_tables, "tonsil_bcell_dars.csv"))
```

```{r}
kbl(head(tonsil_bcell_dars),caption = "Table of the first 5 differentially accesible regions") %>%
  kable_paper("striped", full_width = F)
```


# Save


```{r}
saveRDS(tonsil_bcell,paste0(path_to_save,"3.tonsil_bcell.rds"))
```


<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```


</details>